Firestore/Swift/Source/Codable/SelfDocumentId.swift

```diff
@@ -16,32 +16,37 @@
 
 import FirebaseFirestore
 
-/// A special type used to mark a custom object field to be automatically
-/// populated with the document's referecne when the custom object is created
-/// from a Cloud Firestore document (for example, via
-/// `DocumentSnapshot.data()`).
+/// A value that is populated in Codable objects with a `DocumentReference` by
+/// the FirestoreDecoder when a document is read.
+///
+/// Note that limitations in Swift-compiler generated Codable implementations
+/// prevent using this type wrapped in an Optional. Optional SelfDocumentIDs
+/// are possible if you write a custom init(from: Decoder) method.
 ///
 /// If the field name used for this type conflicts with a read document field,
 /// an error is thrown. For example, if a custom object has a field `firstName`
-/// with type AutoPopulatedDocumentId, and there is a property from the
-/// document named `firstName` as well, an error is thrown when you try to read
-/// the document.
-///
-/// When using a custom object to write to a document, the field with this type
-/// is ignored, which allows writing the object back to any document, even if
-/// it is not the origin of the object.
+/// with type SelfDocumentID, and there is a property from the document named
+/// `firstName` as well, an error is thrown when you try to read the document.
 ///
-/// NOTE: Trying to decode to an `AutoPopulatedDocumentId?` type leads to an
-/// error, this is because compiler generated codable implementations checks for
-/// fields existence for optional types.
+/// When writing a Codable object containing a `SelfDocumentID`, its value is
+/// ignored. This allows you to read a document from one path and write it into
+/// another without adjusting the value here.
 ///
 /// NOTE: Trying to encode/decode this type using encoders/decoders other than
 /// FirestoreEncoder leads to an error.
-public final class AutoPopulatedDocumentId: Equatable, Codable {
-  public let ref: DocumentReference
+public final class SelfDocumentID: Equatable, Codable {
+  public let reference: DocumentReference?
+
+  public var id: String? {
```

Comment:
Given the name and our intent, shouldn't `id` be the first property?

Also, isn't it conventional to put initializers first?

Reply:
The style guide just ask you to make sense..so I shuffled a bit, added `// MARK` comments.

Initializers do come first now.